### PR TITLE
WIP LEARNER-3878 Create entitlement-aware CourseRuns in Publisher

### PR DESCRIPTION
**Pull request by amangano-edx**
Modifies `CreateCourseRunView`, `CreateRunFromDashboardView`, and related templates/forms to support creating Seats from CourseEntitlements associated with the parent course.

JIRA: https://openedx.atlassian.net/browse/LEARNER-3878

Need to add unit tests and determine what to do about CourseSearchForm.

### CODE REVIEW

File: course_discovery/apps/publisher/forms.py [hunk 1]
```diff
@@ -210,7 +210,7 @@ class CourseSearchForm(forms.Form):
     """ Course Type ahead Search Form. """
     course = forms.ModelChoiceField(
         label=_('Find Course'),
-        queryset=Course.objects.all(),
+        queryset=Course.objects.none(),
```
Comment (amangano-edx) [inline course_discovery/apps/publisher/forms.py]: This might not be the correct setting for this, but I don't think that `all` is either, as `all` seems to mean that the form renders with all courses included as options for the select, and also that any user can choose any course when submitting the form. I need to spend some more time tomorrow figuring out what makes sense here.

File: course_discovery/apps/publisher/views.py [hunk 2]
```diff
@@ -665,27 +666,42 @@ def _process_post_request(self, request, parent_course, run_form, seat_form, ctx
...
-        })
+        ctx_overrides = {'run_form': run_form}
+
+        if parent_course.uses_entitlements():
+            try:
+                seat_form = self._bind_seat_form_from_entitlement(parent_course)
+            except (CourseEntitlement.DoesNotExist, CourseEntitlement.MultipleObjectsReturned):
+                messages.error(
+                    request,
+                    _('The certificate configuration for this course is incorrect. Please fix it and try again.')
```
Comment (amangano-edx) [inline course_discovery/apps/publisher/views.py:697]: I'm not sure what to do in this case. This code path should only be hit for courses that have `version` set to `ENTITLEMENT_VERSION` and zero or more than one entitlement.

File: course_discovery/apps/publisher/forms.py [hunk 3]
```diff
@@ -220,6 +220,14 @@ class CourseSearchForm(forms.Form):
         required=True,
     )
 
+    def __init__(self, *args, **kwargs):
+        qs = kwargs.pop('queryset')
+        super(CourseSearchForm, self).__init__(*args, **kwargs)
+
+        if qs:
+            self.fields['course'].queryset = qs
```
Comment (amangano-edx) [inline course_discovery/apps/publisher/forms.py:228]: This will probably change. I need to do some more research to determine what makes the most sense here.

File: course_discovery/apps/publisher/forms.py (hunk 1 above)
Comment (amangano-edx) [inline course_discovery/apps/publisher/forms.py:213]: This will probably change. I need to do some more research to determine what makes the most sense here.

File: course_discovery/static/js/publisher/show-hide-seat-form.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            $seatForm.find('#seatPriceBlock').addClass('hidden');
+            $seatForm.find('#id_price').val('0.0');
+            $seatForm.find('#creditPrice').addClass('hidden');
+            $seatForm.find('#id_credit_price').val('0.0');
+
+            // Re-attach the form
+            $seatForm.insertAfter($courseRunForm);
+        }
+    });
+});
```
Comment (amangano-edx) [inline course_discovery/static/js/publisher/show-hide-seat-form.js:30]: I'm not sure if this is an acceptable approach. It seems to work ok, but it's pretty ugly. Curious what @AlasdairSwan thinks about this.

File: course_discovery/apps/publisher/views.py [hunk 5]
```diff
@@ -698,29 +714,38 @@ def _get_parent_course(self):
...
+        ctx_overrides['course_form'] = course_form
         self.parent_course = course_form.cleaned_data.get('course')
+        if self.parent_course.uses_entitlements():
+            ctx_overrides['hide_seat_form'] = True
+            try:
+                seat_form = self._bind_seat_form_from_entitlement(self.parent_course)
+            except (CourseEntitlement.DoesNotExist, CourseEntitlement.MultipleObjectsReturned):
+                messages.error(
+                    request,
+                    _('The certificate configuration for this course is incorrect. Please fix it and try again.')
```
Comment (amangano-edx) [inline course_discovery/apps/publisher/views.py:743]: I'm not sure what to do in this case. This code path should only be hit for courses that have `version` set to `ENTITLEMENT_VERSION` and zero or more than one entitlement.

File: course_discovery/static/js/publisher/show-hide-seat-form.js (hunk 4 above)
Comment (AlasdairSwan) [reply course_discovery/static/js/publisher/show-hide-seat-form.js:30]: @amangano-edx it is pretty ugly, but then it's not using any JS Frameworks so you're limited in your approach I guess. Is this up in a test environment? I think I need some more details to consider alternative approaches? If not I can swing by your desk at some point today.

File: course_discovery/apps/publisher/models.py [hunk 6]
```diff
@@ -100,6 +100,12 @@ class Course(TimeStampedModel, ChangedByMixin):
     def __str__(self):
         return self.title
 
+    def uses_entitlements(self):
```
Comment (mikix) [inline course_discovery/apps/publisher/models.py:103]: Might be nicer as a @property, but fine as-is too.

File: course_discovery/static/js/publisher/show-hide-seat-form.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    // If the rendered form is hidden, remove it from the DOM.
+    if ($seatForm.hasClass('hidden')) {
+        $seatForm.detach();
+        $seatForm.removeClass('hidden');
+    }
+
+    $(document).on('select2:select', '#id_course', function(e) {
+        var usesEntitlements = e.params.data.uses_entitlements;
+
+        $seatForm.detach();
```
Comment (AlasdairSwan) [inline course_discovery/static/js/publisher/show-hide-seat-form.js:14]: Why detach here, if ```!usesEntitlements``` you don't need to detach do you?

Comment (amangano-edx) [reply course_discovery/static/js/publisher/show-hide-seat-form.js:14]: That's true.

Comment (amangano-edx) [reply course_discovery/static/js/publisher/show-hide-seat-form.js:14]: You need to detach before re-attaching to ensure that the form is accidentally added twice.

**Comment by amangano-edx** [comment]
Replaced by https://github.com/edx/course-discovery/pull/1306
